lib/temporal/worker/activity_worker.rb

```diff
@@ -0,0 +1,103 @@
+require 'temporal/error/failure'
+require 'temporal/worker/sync_worker'
+require 'temporal/worker/activity_runner'
+
+module Temporal
+  class Worker
+    class ActivityWorker
+      def initialize(task_queue, core_worker, activities, converter, executor)
+        @running = false
+        @task_queue = task_queue
+        @worker = SyncWorker.new(core_worker)
+        @activities = prepare_activities(activities)
+        @converter = converter
+        @executor = executor
+        @running_activities = {}
+      end
+
+      def run(reactor)
+        @running = true
+
+        while running?
+          activity_task = worker.poll_activity_task
+          reactor.async do
+            if activity_task.start
+              handle_start_activity(activity_task.task_token, activity_task.start)
+            elsif activity_task.cancel
+              handle_cancel_activity(activity_task.task_token, activity_task.cancel)
+            end
+          end
+        end
+      end
+
+      def shutdown
+        @running = false
+        executor.shutdown
+        # TODO: core worker shutdown implementation pending
+      end
+
+      private
+
+      attr_reader :task_queue, :worker, :activities, :converter, :executor, :running_activities
+
+      def running?
+        @running
+      end
+
+      def prepare_activities(activities)
+        activities.each_with_object({}) do |activity, result|
+          unless activity.ancestors.include?(Temporal::Activity)
+            raise ArgumentError, 'Activity must be a subclass of Temporal::Activity'
+          end
+
+          if result[activity._name]
+            raise ArgumentError, "More than one activity named #{activity._name}"
+          end
+
+          result[activity._name] = activity
+          result
+        end
+      end
+
+      def lookup_activity(activity_type)
+        activities.fetch(activity_type) do
+          activity_names = activities.keys.sort.join(', ')
+          raise Temporal::Error::ApplicationError.new(
+            "Activity function #{activity_type} is not registered on this worker, " \
```

Comment:
Not sure you want to call this `function` since activities are dispatched based on their class name.

Reply:
Ah, my bad, copied error from another SDK